File: fast_bocpd/hazard.py
import math


class ConstantHazard:
    """
    Constant hazard model: H = 1 / lambda_.

    At each time step, independent of the current run length r:
        P(changepoint)  = H
        P(continuation) = 1 - H
    """
    def __init__(self, lambda_: float):
        if lambda_ <= 0:
            raise ValueError("lambda_ must be > 0")
        self.lambda_ = float(lambda_)
        self.H = 1.0 / self.lambda_
        if not (0.0 < self.H < 1.0):
            # For extreme lambda_, this could under/overflow numerically.
            raise ValueError("Hazard H = 1/lambda_ must be in (0, 1).")

        self.log_H = math.log(self.H)
        self.log_1mH = math.log(1.0 - self.H)

    def log_transition_cp(self, r_prev: int) -> float:
        """
        log P(r_t = 0 | r_{t-1} = r_prev)

        For a constant hazard, this is always log(H).
        """
        return self.log_H

    def log_transition_cont(self, r_prev: int) -> float:
        """
        log P(r_t = r_prev + 1 | r_{t-1} = r_prev)

        For a constant hazard, this is always log(1 - H).
        """
        return self.log_1mH

File: fast_bocpd/_c/hazard.c
#include "hazard.h"
#include <math.h>

int constant_hazard_init(ConstantHazardParams* params, double lambda) 
{
    if (lambda <= 0.0) {
        return -1;  // Invalid lambda
    }

    params->lambda = lambda;
    double H = 1.0 / lambda;
    
    if (H <= 0.0 || H >= 1.0) {
        return -1;  // Hazard must be in (0, 1)
    }

    params->log_H = log(H);
    params->log_1mH = log(1.0 - H);
    
    return 0;
}

double constant_hazard_log_transition_cp(const ConstantHazardParams* params, int32_t r_prev)
{
    // For constant hazard, changepoint probability is always log(H)
    // r_prev is unused
    (void)r_prev;  // Suppress unused parameter warning
    return params->log_H;
}

double constant_hazard_log_transition_cont(const ConstantHazardParams* params, int32_t r_prev)
{
    // For constant hazard, continuation probability is always log(1 - H)
    // r_prev is unused
    (void)r_prev;  // Suppress unused parameter warning
    return params->log_1mH;
}
